### ISSUE TITLE
Regression in type inference for a union of array types inside an object type

### ISSUE THREAD

**Issue by daitay4** [colinhacks/zod#2203]
I have some schema like the following:
```
z.object({
    arrayProp: z.union([z.string().array(), z.number().array()]),
}),
```

this used to produce the following type for the `arrayProp` key:  `string[] | number[]`.

since version 3.21.2 this produces the following type: 
`string[] | (string[] & number[]) | (string[] & undefined) | (number[] & string[]) | number[] | (number[] & undefined)`

this in turn causes many problems when trying to work with `arrayProp`, for example if I try to use an array predicate like `map`, I'm getting the following error: 
```
TS2349: This expression is not callable.   Each member of the union type '(<U>(callbackfn: (value: string, index: number, array: string[]) => U, thisArg?: any) => U[]) | (<U>(callbackfn: (value: number, index: number, array: number[]) => U, thisArg?: any) => U[]) | ((<U>(callbackfn: (value: string, index: number, array: string[]) => U, thisArg?: any) => U[]) & (<U>(callbackfn: (value: num...' has signatures, but none of those signatures are compatible with each other.
```

However, if I remove the wrapper `z.object` and use the following schema: `z.union([z.string().array(), z.number().array()])` then I'm getting the correct type inference of `string[] | number[]`

Looking at the code for the 3.21.2 PR (https://github.com/colinhacks/zod/commit/4d016b772b79d566bfa2a0c2fc5bfbd92b776982#diff-a5a27c41f819170124f4fd4cf24b1d71b8ab2966a4bde235a7f8db53dc8ef486) suggests that this has to do with changes made to the internal`DeepPartial` type.

For now we're sticking with 3.21.1 version, but I would be happy to see that fixed as we use this pattern on multiple occasions across our project.

**Comment by jer-sen** [comment]
I face something similar:
```typescript
const prop = z.string().or(z.object({}));
const obj = z.object({ prop });
type ObjectPropType = z.TypeOf<typeof obj>['prop'];
// type ObjectPropType = (string | {}) & (string | {} | undefined)
type PropType = z.TypeOf<typeof prop>;
// type PropType = string | {}
```

**Comment by Michal-Miky-Jankovsky** [comment]
Maybe the problem is in the typescript itself.

https://github.com/microsoft/TypeScript/issues/44373

**Comment by dcfreire** [comment]
Interestingly calling `.partial()` on a `z.object` produces the correct type.

**Comment by eyalch** [comment]
Bump.

As mentioned in the original description, this issue forces us to stay at version 3.21.1.

**Comment by xsjcTony** [comment]
Any updates? This is definitely something need to be fixed

**Comment by xsjcTony** [comment]
a simple one
```ts
const schema = z.union([
  z.union([z.string(), z.number()]),
  z.union([z.string(), z.number()]).array()
]);
```
ends up
```
description: string | (string & (string | number)[]) | number | (number & (string | number)[]) | ((string | number)[] & string) | ((string | number)[] & number) | (string | number)[] | ((string | number)[] & undefined)
```

**Comment by jon-dez** [comment]
Simpler one:

![image](https://github.com/colinhacks/zod/assets/48642222/444e8a6c-017d-4595-adf8-ad743584102d)

**Comment by Joeri1324** [comment]
Is there any update on this issue? I'm running the exact same problem.

**Comment by khusmann** [comment]
It also happens with tuples as well:

```
const foo = z.tuple([z.number(), z.number()]).or(z.number());

type Foo = z.infer<typeof foo>;
// [number, number] | number

const bar = z.object({ foo });

type Bar = z.infer<typeof bar>;
// { foo: (number | [number, number]) & (number | [number, number]) | undefined } ??
```

**Comment by KubaJastrz** [comment]
Stumbled upon this in latest version 3.22.4 
https://www.typescriptlang.org/play#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFBkDGEAdgM7wCuNwtcAvCgHTOs0AUAbWRcGUYDQDm-AJRcAhlCjyAnrIA0ImoxAAjAKZRZCpatkBdGRRgqw+rABt5MAPo27HbhMyGAPO-0ITDheWgA+MgB6SLhYgD0AfgpqengafQZ9Yhc6SgALfRB5TxEIXQArfUoYfiRQmnQrMgC4dMzsls4Rbz8AoNaMmCyc-ML5MMF8evxzKJj4pKA

**Comment by cyc1e** [comment]
Any updates?

**Comment by fdc-viktor-luft** [comment]
Since it seems like no one is able to fix it, I wanted to share at least a "hacky" workaround for the inferred type.

```ts
const bSchema = z.union([z.object({ c: z.string() }), z.object({ d: z.string() })]);
const FooSchema = z.object({
    a: z.string(),
    b: bSchema,
});

export type Foo = z.infer<typeof FooSchema>;
```

Produces the undesired type:

```ts
type Foo = {
    a: string;
    b: ({
        c: string;
    } | {
        d: string;
    }) & ({
        c: string;
    } | {
        d: string;
    } | undefined);
}
```

You can do:

```ts
export type Foo = Omit<z.infer<typeof FooSchema>, "b"> & { b: z.infer<typeof bSchema> };
```

To get:

```ts
type Foo = {
    a: string;
    b: {
        c: string;
    } | {
        d: string;
    };
}
```

But this is of course far from ideal. I tried shortly to look in the Zod codebase to find the root cause, but - oh well - that is really complex and I have too less time to dig into it. Maybe @mattpocock who promotes this library so much has some time fixing it? 😅

**Comment by homemade-ramen** [comment]
It appears that this problem may be fixed in the latest version (v3.23.0 and up). The potential fix is included in https://github.com/colinhacks/zod/pull/3352

You can check if the issue is fixed in the [TypeScript Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgFBkDGEAdgM7yYQRwC8KAdBAEYBWAppRgAKBGTgS4AQyhQpATwAKOMAC5OAVxrBawgNrIODKMBoBzYQEoOMufKsAaTjQ0hu-KFZuyFVgLqWDmRolhQw8mD8cABizGycppgeADzhkRCYWMwAfGQA9HmSAHoA-GRAA).

```typescript
import { z } from 'zod'

const foo = z.object({
    arrayProp: z.union([z.string().array(), z.number().array()]),
})

type Foo = z.infer<typeof foo>
//   ^?  type Foo = {arrayProp: string[] | number[]}
```

Additionally, https://github.com/colinhacks/zod/issues/2325 can also be fixed with the same fix.

@colinhacks, could you please review this and confirm that the issues have been resolved?

**Comment by fdc-viktor-luft** [comment]
I can confirm that my above posted issue was resolved with the most recent version.

**Comment by daitay4** [comment]
Closing this issue as it seems to have been resolved, perhaps via #3352 as noted by @homemade-ramen
